File: 0022.GenerateParentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        if n < 1:
            return
        ans = []
        #self.dfs(2*n, n, n, '', ans)
        self.dfs2(n, n, '', ans)
        return ans

    def dfs(self, n, left, right, path, ans):
        if left > right:
            return
        if n == 0:
            ans.append(path)
            return
        if left > 0:
            self.dfs(n-1, left-1, right, path+'(', ans)
        if right > 0:
            self.dfs(n-1, left, right-1, path+')', ans)

    def dfs2(self, left, right, path, ans):
        if left > right:
            return
        if left == 0 and right == 0:
            ans.append(path)
            return
        if left > 0:
            self.dfs2(left-1, right, path+'(', ans)
        if right > 0:
            self.dfs2(left, right-1, path+')', ans)

File: 0022.GenerateParentheses.go
func generateParenthesis(n int) []string {
	if n == 0 {
		return []string{""}
	}
	ans := []string{}
	//dfs(n, n, "", &ans)
	for i := 0; i < n; i++ {
		for _, left := range generateParenthesis(i) {
			for _, right := range generateParenthesis(n - i - 1) {
				ans = append(ans, "("+left+")"+right)
			}
		}
	}
	return ans
}

func dfs(left, right int, path string, ans *[]string) {
	if left > right || left < 0 || right < 0 {
		return
	}
	if left == 0 && right == 0 {
		*ans = append(*ans, path)
	}
	dfs(left-1, right, path+"(", ans)
	dfs(left, right-1, path+")", ans)
}